substrate/emissions/pallet/src/lib.rs

```diff
@@ -0,0 +1,413 @@
+    /// Networks to spawn Serai with.
+    pub networks: Vec<(NetworkId, Amount)>,
+    /// List of participants to place in the initial validator sets.
+    pub participants: Vec<T::AccountId>,
+  }
+
+  impl<T: Config> Default for GenesisConfig<T> {
+    fn default() -> Self {
+      GenesisConfig { networks: Default::default(), participants: Default::default() }
+    }
+  }
+
+  #[pallet::error]
+  pub enum Error<T> {
+    NetworkHasEconomicSecurity,
+    NoValueForCoin,
+  }
+
+  #[pallet::event]
+  pub enum Event<T: Config> {}
+
+  #[pallet::pallet]
+  pub struct Pallet<T>(PhantomData<T>);
+
+  #[pallet::storage]
+  #[pallet::getter(fn participants)]
+  pub(crate) type Participants<T: Config> = StorageMap<
+    _,
+    Identity,
+    NetworkId,
+    BoundedVec<(PublicKey, u64), ConstU32<{ MAX_KEY_SHARES_PER_SET }>>,
+    OptionQuery,
+  >;
+
+  #[pallet::storage]
+  #[pallet::getter(fn session)]
+  pub type CurrentSession<T: Config> = StorageMap<_, Identity, NetworkId, u32, ValueQuery>;
+
+  #[pallet::storage]
+  #[pallet::getter(fn economic_security_reached)]
+  pub(crate) type EconomicSecurityReached<T: Config> =
+    StorageMap<_, Identity, NetworkId, bool, ValueQuery>;
+
+  #[pallet::storage]
+  #[pallet::getter(fn last_swap_volume)]
+  pub(crate) type LastSwapVolume<T: Config> = StorageMap<_, Identity, NetworkId, u64, OptionQuery>;
+
+  #[pallet::storage]
+  pub(crate) type GenesisCompleteBlock<T: Config> = StorageValue<_, u64, OptionQuery>;
+
+  #[pallet::genesis_build]
+  impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+    fn build(&self) {
+      for (id, stake) in self.networks.clone() {
+        let mut participants = vec![];
+        for p in self.participants.clone() {
+          participants.push((p, stake.0));
+        }
+        Participants::<T>::set(id, Some(participants.try_into().unwrap()));
+        CurrentSession::<T>::set(id, 0);
+        EconomicSecurityReached::<T>::set(id, false);
+      }
+    }
+  }
+
+  #[pallet::hooks]
+  impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+    fn on_initialize(n: BlockNumberFor<T>) -> Weight {
+      if GenesisCompleteBlock::<T>::get().is_none() &&
+        GenesisLiquidity::<T>::genesis_complete().is_some()
+      {
+        GenesisCompleteBlock::<T>::set(Some(n.saturated_into::<u64>()));
+      }
+
+      // we wait 1 extra block after genesis ended to see the changes. We only need this extra
+      // block in dev&test networks where we start the chain with accounts that already has some
+      // staked SRI. So when we check for ec-security pre-genesis we look like we are economically
+      // secure. The reason for this although we only check for it once the genesis is complete(so
+      // if the genesis complete we shouldn't be economically secure because the funds are not
+      // enough) is because ValidatorSets pallet runs before the genesis pallet in runtime.
+      //  So ValidatorSets pallet sees the old state until next block.
```

Comment:
Can we simply have the ValidatorsSets pallet run after? What updates does the ValidatorSets pallet do that's relevant here? Since we have to program in genesis validators not actually having stake at some point, so we should not there here and schedule this to be removed upon that `TODO` being completed?

Reply:
I'm fine with the re-ordering. I just am trying work towards minimize these weird edge cases which require paragraphs explaining them. Some will always be justified.

> Since we have to program in genesis validators not actually having stake at some point

The genesis validators for mainnet will not have stake. They will be considered to have stake for the purpose of validator selection. Their only stake will be the rewards they earn.

> so we should not there here and schedule this to be removed upon that TODO being completed?

> > so we should not there here

*so we should note this here

> > and schedule this to be removed upon that TODO being completed

Since eventually, genesis validators will not actually have stake per EconSec considerations, the necessity for this workaround should fall away.